File: android_app/src/main/java/me/timeto/app/ui/ChecklistsPickerSheet.kt
package me.timeto.app.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.unit.dp
import me.timeto.app.*
import me.timeto.shared.db.ChecklistDb
import me.timeto.shared.vm.Checklist.ChecklistPickerVm

@Composable
fun ChecklistsPickerSheet(
    layer: WrapperView.Layer,
    selectedChecklists: List<ChecklistDb>,
    onPick: (List<ChecklistDb>) -> Unit,
) {

    val (vm, state) = rememberVm { ChecklistPickerVm(selectedChecklists) }

    VStack(
        modifier = Modifier
            .fillMaxHeight()
            .background(c.sheetBg)
    ) {

        val scrollState = rememberScrollState()

        Sheet.HeaderViewOld(
            onCancel = { layer.close() },
            title = state.headerTitle,
            doneText = state.doneTitle,
            isDoneEnabled = true,
            scrollState = scrollState,
        ) {
            onPick(vm.getSelectedChecklists())
            layer.close()
        }

        VStack(
            modifier = Modifier
                .verticalScroll(state = scrollState)
                .padding(bottom = 20.dp)
                .navigationBarsPadding()
                .imePadding()
        ) {

            ZStack(Modifier.height(20.dp)) { }

            val checklistsUI = state.checklistsUI
            checklistsUI.forEach { checklistUI ->
                val isFirst = state.checklistsUI.first() == checklistUI
                MyListView__ItemView(
                    isFirst = isFirst,
                    isLast = state.checklistsUI.last() == checklistUI,
                    withTopDivider = !isFirst,
                ) {
                    MyListView__ItemView__RadioView(
                        text = checklistUI.text,
                        isActive = checklistUI.isSelected,
                    ) {
                        vm.toggleChecklist(checklistUI)
                    }
                }
            }

            Text(
                text = state.newChecklistButton,
                modifier = Modifier
                    .padding(top = 12.dp, start = 14.dp)
                    .clip(squircleShape)
                    .clickable {
                        Dialog.show { layer ->
                            ChecklistNameDialog(
                                layer = layer,
                                editedChecklist = null,
                                onSave = { newChecklistDb ->
                                    vm.selectById(newChecklistDb.id)
                                    Sheet.show { layer ->
                                        ChecklistFormSheet(
                                            layer = layer,
                                            checklistDb = newChecklistDb,
                                            onDelete = {},
                                        )
                                    }
                                },
                            )
                        }
                    }
                    .padding(horizontal = 8.dp, vertical = 4.dp),
                color = c.blue,
            )

            ZStack(Modifier.height(20.dp)) { }
        }
    }
}

File: ios_app/iphone/ChecklistsPickerSheet.swift
import SwiftUI
import shared

struct ChecklistsPickerSheet: View {

    @Binding private var isPresented: Bool
    private let onPick: ([ChecklistDb]) -> Void

    @State private var vm: ChecklistsPickerSheetVm
    @State private var sheetHeaderScroll = 0

    @EnvironmentObject private var nativeSheet: NativeSheet

    init(
        isPresented: Binding<Bool>,
        selectedChecklists: [ChecklistDb],
        onPick: @escaping ([ChecklistDb]) -> Void
    ) {
        self.onPick = onPick
        _isPresented = isPresented
        _vm = State(initialValue: ChecklistsPickerSheetVm(selectedChecklists: selectedChecklists))
    }

    var body: some View {

        VMView(vm: vm, stack: .VStack()) { state in

            SheetHeaderView(
                onCancel: { isPresented = false },
                title: state.headerTitle,
                doneText: state.doneTitle,
                isDoneEnabled: true,
                scrollToHeader: sheetHeaderScroll
            ) {
                onPick(vm.getSelectedChecklists())
                isPresented = false
            }

            ScrollViewWithVListener(showsIndicators: false, vScroll: $sheetHeaderScroll) {

                VStack {

                    let checklistsUI = state.checklistsUI
                    ForEach(checklistsUI, id: \.checklist.id) { checklistUI in

                        let isFirst = checklistsUI.first == checklistUI

                        MyListView__ItemView(
                            isFirst: isFirst,
                            isLast: checklistsUI.last == checklistUI,
                            withTopDivider: !isFirst
                        ) {

                            MyListView__ItemView__RadioView(
                                text: checklistUI.text,
                                isActive: checklistUI.isSelected
                            ) {
                                vm.toggleChecklist(checklistUI: checklistUI)
                            }
                        }
                    }

                    HStack {

                        Button(
                            action: {
                                nativeSheet.show { isPresentedNameDialog in
                                    ChecklistNameDialog(
                                        isPresented: isPresentedNameDialog,
                                        checklist: nil,
                                        onSave: { newChecklistDb in
                                            vm.selectById(id: newChecklistDb.id)
                                            nativeSheet.show { isPresentedForm in
                                                ChecklistFormSheet(
                                                    checklistDb: newChecklistDb,
                                                    onDelete: {}
                                                )
                                            }
                                        }
                                    )
                                }
                            },
                            label: {
                                Text(state.newChecklistButton)
                                    .foregroundColor(c.blue)
                                    .padding(.leading, H_PADDING + 6)
                                    .padding(.top, 16)
                            }
                        )

                        Spacer()
                    }
                }
                .padding(.top, 20)
            }
        }
        .background(c.sheetBg)
    }
}
